### PR TITLE
[MPORT-501] Include secure settings check in validations

### PR DESCRIPTION
**Pull request by keitheous**
### Descriptions
We are adding a new **validation warning (non-blocking) to suggest** using secure settings IF developers have token or passwords as a parameter.

**Steps:**
1) Look at manifest parameter names
2) If any of them has the word "token", "key", or "password" in them, display this message.

    image
    Link: https://developer.zendesk.com/apps/docs/developer-guide/using_sdk#using-secure-settings

3) Update other strings that are not related to this PR.

**Todos:**
- [ ] Use translated strings before merging
- [x] Add spec
- [x] Clean up Commits from the 3 to 2 
- [x] Make sure [this](https://github.com/zendesk/zendesk_apps_support/pull/235) is merged first. 

### References
- https://zendesk.atlassian.net/browse/MPORT-501

### Risks
[low] Validation WARNING on secure settings doesn't work properly. 

[MPORT-501]: https://zendesk.atlassian.net/browse/MPORT-501

### CODE REVIEW

File: lib/zendesk_apps_support/validations/secure_settings.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module ZendeskAppsSupport
+  module Validations
+    module SecureSettings
+      SECURABLE_KEYWORDS = %w[token key pwd password].freeze
+      SECURABLE_KEYWORDS_REGEXP = /#{ Regexp.union(SECURABLE_KEYWORDS).source }/i
```
Comment (seangoedecke) [inline lib/zendesk_apps_support/validations/secure_settings.rb]: I'm sad that there's no nicer way to do this than to compile a regex, unwrap it into a string, then compile that into another regex. I guess you could do something like `Regexp.new(SECURABLE_KEYWORDS.join('|'), Regexp::IGNORECASE)` - not sure if that's nicer.

File: lib/zendesk_apps_support/validations/secure_settings.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+          manifest_params = package.manifest.parameters
+
+          insecure_params_found = manifest_params.any? { |param| insecure_param(param) }
+
+          package.warnings << secure_settings_warning if insecure_params_found
+        end
+
+        private
+
+        def insecure_param(parameter)
```
[the remark this answers is not in the document]
Comment (keitheous) [reply lib/zendesk_apps_support/validations/secure_settings.rb]: Yeah I did think about that. The first part of this equation does not return a boolean
```
parameter.name =~ SECURABLE_KEYWORDS_REGEXP && type_password_or_text?(parameter.type) && !parameter.secure
```
I tried`!!(parameter.name =~ SECURABLE_KEYWORDS_REGEXP)` but rubocop doesn't sound happy.
Tried `.match?` also, but some ruby versions do not know this method. Hence the missing `?`

However, now that I think about it, `&&` to any truthy value would force it into a boolean. Very happy to change that and add the `?` 👍

File: lib/zendesk_apps_support/package.rb [hunk 3]
```diff
@@ -39,6 +39,9 @@ def validate(marketplace: true, skip_marketplace_translations: false)
         errors << Validations::Requirements.call(self)
         errors << Validations::Requests.call(self)
 
+        # warning only validators with valid manifest
```
[the remark this answers is not in the document]
Comment (keitheous) [reply lib/zendesk_apps_support/package.rb]: I'll explain. Further down in this file, there is another comment that says
```
# warning only validators
Validations::Secrets.call(self)
```

warning only validators appends to `package.warnings` instead of `errors`. Errors blocks app uploads where as warnings do not. They are printed as outputs.

Initially I wanted to put `Validations::SecureSettings.call(self)` together with `Validations::Secrets.call(self)`.

After a while I thought that I should only check for this if they have a valid manifest.

[the remark this answers is not in the document]
Comment (keitheous) [reply lib/zendesk_apps_support/package.rb]: Is there a clearer way to do this? Should I just remove the comment?

[the remark this answers is not in the document]
Comment (seangoedecke) [reply lib/zendesk_apps_support/package.rb]: I'd reword the comment to something like `# only adds warnings` in both places - the relevant thing to explain here is why you're not doing `errors <<`, so there's no need to mention anything about the manifest IMO

File: lib/zendesk_apps_support/validations/secure_settings.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (seangoedecke) [reply lib/zendesk_apps_support/validations/secure_settings.rb]: Hmm, interesting. I still think it's fine to have a `?` method even if it's not returning a boolean, as long as you only care about whether the method return value is truthy

File: lib/zendesk_apps_support/package.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (keitheous) [reply lib/zendesk_apps_support/package.rb]: 👍 Gotcha

File: lib/zendesk_apps_support/validations/secure_settings.rb (hunk 1 above)
Comment (keitheous) [reply lib/zendesk_apps_support/validations/secure_settings.rb]: Sounds good to me!
